src/controllers/rolesController.js

```diff
@@ -1,24 +1,23 @@
-const UserProfile = require("../models/userProfile");
-const cache = require("../utilities/nodeCache")();
-const { hasPermission } = require("../utilities/permissions");
+const UserProfile = require('../models/userProfile');
```

Comment:
Great job on the recent code! It shows a solid understanding of Node.js and Express for handling roles and permissions. Here are a few suggestions that could further enhance the readability, maintainability, and performance of your code:
1. Improve Error Handling:

Issue: Error handling could be more standardized and descriptive.
Suggestion: Consider using a centralized error handling middleware to maintain consistency across your application and provide more detailed error messages.

```
const errorHandler = (err, req, res, next) => {
  res.status(err.status || 500).send({ error: err.message || 'Internal Server Error' });
};
```
2. Consistent Asynchronous Code:

Issue: Mixing callback-based and async/await-based code can lead to inconsistencies.
Suggestion: Refactor your code to consistently use async/await.

```
const getAllRoles = async (req, res, next) => {
  try {
    const results = await Role.find({});
    res.status(200).send(results);
  } catch (error) {
    next(error);
  }
};

const createNewRole = async (req, res, next) => {
  try {
    const newRole = new Role({
      roleName: req.body.roleName,
      permissions: req.body.permissions,
    });
    const savedRole = await newRole.save();
    res.status(201).send(savedRole);
  } catch (error) {
    next(error);
  }
};
```
Enhanced Input Validation:

Issue: Current input validation is minimal and could be more robust.
Suggestion: Utilize a validation library like Joi or express-validator for more comprehensive input validation.

Summary of Changes:
Centralized Error Handling: Adds middleware for consistent error handling.
Consistent Async/Await Usage: Refactors functions to use async/await.
Permission Check Middleware: Moves permission checks to middleware.
Enhanced Input Validation: Adds middleware for more robust input validation.
These improvements will make your codebase more modular, readable, and easier to maintain. Keep up the fantastic work!